cmd/thanos/rule.go

```diff
@@ -738,36 +719,49 @@ func queryFunc(
 	}
 
 	return func(ctx context.Context, q string, t time.Time) (promql.Vector, error) {
-		// Add DNS resolved addresses from static flags and file SD.
-		// TODO(bwplotka): Consider generating addresses in *url.URL.
-		addrs := dnsProvider.Addresses()
-
-		removeDuplicateQueryAddrs(logger, duplicatedQuery, addrs)
-
-		for _, i := range rand.Perm(len(addrs)) {
-			u, err := url.Parse(fmt.Sprintf("http://%s", addrs[i]))
-			if err != nil {
-				return nil, errors.Wrapf(err, "url parse %s", addrs[i])
-			}
-
-			span, ctx := tracing.StartSpan(ctx, spanID)
-			v, warns, err := promclient.PromqlQueryInstant(ctx, logger, u, q, t, promclient.QueryOptions{
-				Deduplicate:             true,
-				PartialResponseStrategy: partialResponseStrategy,
-			})
-			span.Finish()
+		for _, i := range rand.Perm(len(queriers)) {
+			querier := queriers[i]
+			c := promclient.NewClient(logger, querier)
+			endpoints := removeDuplicateQueryEndpoints(logger, duplicatedQuery, querier.Endpoints())
+			for _, i := range rand.Perm(len(endpoints)) {
+				span, ctx := tracing.StartSpan(ctx, spanID)
+				v, warns, err := c.PromqlQueryInstant(ctx, endpoints[i], q, t, promclient.QueryOptions{
+					Deduplicate:             true,
+					PartialResponseStrategy: partialResponseStrategy,
+				})
+				span.Finish()
 
-			if err != nil {
-				level.Error(logger).Log("err", err, "query", q)
-			} else {
-				if len(warns) > 0 {
-					ruleEvalWarnings.WithLabelValues(strings.ToLower(partialResponseStrategy.String())).Inc()
-					// TODO(bwplotka): Propagate those to UI, probably requires changing rule manager code ):
-					level.Warn(logger).Log("warnings", strings.Join(warns, ", "), "query", q)
+				if err != nil {
+					level.Error(logger).Log("err", err, "query", q)
+				} else {
+					if len(warns) > 0 {
+						ruleEvalWarnings.WithLabelValues(strings.ToLower(partialResponseStrategy.String())).Inc()
+						// TODO(bwplotka): Propagate those to UI, probably requires changing rule manager code ):
+						level.Warn(logger).Log("warnings", strings.Join(warns, ", "), "query", q)
+					}
+					return v, nil
 				}
-				return v, nil
 			}
 		}
 		return nil, errors.Errorf("no query peer reachable")
 	}
 }
+
+func addToGroup(g *run.Group, c *http_util.FanoutClient, interval time.Duration) {
```

Comment:
It's not only add, it adds for DNS & file discovery, so maybe: 

```suggestion
func addDiscoveryGroups(g *run.Group, c *http_util.FanoutClient, interval time.Duration) {
```